### PR TITLE
reconstruction api refactor for frag

### PR DESCRIPTION
**Pull request by assafvayner**

### CODE REVIEW

File: cas_client/src/remote_client.rs [hunk 1]
```diff
@@ -204,29 +206,61 @@ impl RemoteClient {
...
-                .await
-                .map_err(|e| CasClientError::InternalError(anyhow!("join error {e}")))??;
-            writer.write_all(&piece)?;
+        let terms = reconstruction_response.terms;
+        let fetch_info = reconstruction_response.fetch_info;
+        let total_len = terms.iter().fold(0, |acc, x| acc + x.unpacked_length);
+
+        let mut cache: HashMap<&CASReconstructionFetchInfo, (Vec<u32>, Vec<u8>)> = HashMap::new();
+        for term in terms.iter() {
+            let fhash = fetch_info
```
Comment (bpronan) [inline cas_client/src/remote_client.rs:215]: Minor nit: I would prefer these were named `fetch_hash` and `fetch_term` to improve readability.

File: cas_object/src/cas_chunk_format.rs [hunk 2]
```diff
@@ -165,6 +165,26 @@ pub fn deserialize_chunk_to_writer<R: Read, W: Write>(
     Ok(header.get_compressed_length() as usize + CAS_CHUNK_HEADER_LENGTH)
 }
 
+pub fn deserialize_chunks2<R: Read>(reader: &mut R) -> Result<(Vec<u32>, Vec<u8>), CasObjectError> {
```
Comment (bpronan) [inline cas_object/src/cas_chunk_format.rs:168]: Do we need both functions (`deserialize_chunks` and `deserialize_chunks2`) around? If we are keeping both around, let's rename this one to something like `deserialize_indexed_chunks`.

File: cas_object/src/cas_chunk_format.rs [hunk 3]
```diff
@@ -165,6 +165,26 @@ pub fn deserialize_chunk_to_writer<R: Read, W: Write>(
...
+    let mut buf = Vec::new();
+    let mut indices = vec![0];
+    loop {
```
Comment (bpronan) [inline cas_object/src/cas_chunk_format.rs:171]: Not a deal breaker for this check-in, but I don't like that this is a effectively hidden a 2x nested "infinite" loop. Can we collapse both functions into a single loop? 

Non-conditional loops are troublesome enough on their own merit.

File: cas_object/src/cas_chunk_format.rs (hunk 2 above)
Comment (assafvayner) [reply cas_object/src/cas_chunk_format.rs:168]: don't need this, Rajat modified the correct one to provide this.

**Pull request closed without merging**